crates/rpc/rpc-types/src/eth/filter.rs

```diff
@@ -9,13 +10,100 @@ use serde::{
     ser::SerializeStruct,
     Deserialize, Deserializer, Serialize, Serializer,
 };
-use std::ops::{Range, RangeFrom, RangeTo};
+use std::{
+    collections::HashSet,
+    hash::Hash,
+    ops::{Range, RangeFrom, RangeTo},
+};
 
 /// Helper type to represent a bloom filter used for matching logs.
-pub type BloomFilter = Vec<Option<Bloom>>;
+#[derive(Default, Debug)]
+pub struct BloomFilter(Vec<Bloom>);
+
+impl From<Vec<Bloom>> for BloomFilter {
+    fn from(src: Vec<Bloom>) -> Self {
+        BloomFilter(src)
+    }
+}
+
+impl BloomFilter {
+    /// Returns whether the given bloom matches the list of Blooms in the current filter.
+    /// If the filter is empty (the list is empty), then any bloom matches
+    /// Otherwise, there must be at least one matche for the BloomFilter to match.
+    pub fn matches(&self, bloom: Bloom) -> bool {
+        self.0.is_empty() || self.0.iter().any(|a| bloom.contains_bloom(a))
+    }
+}
+
+#[derive(Default, Debug, PartialEq, Eq, Clone, Deserialize)]
+/// FilterSet is a set of values that will be used to filter logs
+pub struct FilterSet<T: Eq + Hash>(HashSet<T>);
```

Comment:
Yep, only drawback is having a worst inclusion check. This could be balanced by having a custom serialization that orders the `Vec` in ascending order? I'm quite sure how this would be an issue though?